Core/clim-core/redefinition.lisp

```diff
@@ -0,0 +1,177 @@
+;;;; (C) Copyright 2019, 2020 Jan Moringen
+;;;;
+;;;; This library is free software; you can redistribute it and/or
+;;;; modify it under the terms of the GNU Library General Public
+;;;; License as published by the Free Software Foundation; either
+;;;; version 2 of the License, or (at your option) any later version.
+;;;;
+;;;; This library is distributed in the hope that it will be useful,
+;;;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+;;;; Library General Public License for more details.
+;;;;
+;;;; You should have received a copy of the GNU Library General Public
+;;;; License along with this library; if not, write to the
+;;;; Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+;;;; Boston, MA  02111-1307  USA.
+
+(in-package #:clim-internals)
+
+;;; Frame class redefinition notification protocol
+;;;
+;;; This is used for notifying application frame classes after aspects
+;;; such as the class itself, the `generate-panes' method or the
+;;; command table have been redefined. The notification sequence
+;;; begins with a call to `note-frame-class-redefined' which either
+;;; directly calls `handle-frame-class-redefinition' or schedules the
+;;; call to be made when leaving a
+;;; `with-delayed-redefinition-notifications'.
+;;;
+;;; This protocol is probably most commonly initiated by re-evaluating
+;;; the `define-application-frame' form that defined the frame class
+;;; in question, but any event that causes the `reinitialize-instance'
+;;; method of the class to execute also triggers it. In the context of
+;;; `define-application-frame',
+;;; `with-delayed-redefinition-notifications' delays the
+;;; `note-frame-class-redefinition' call until after all aspects
+;;; (`generate-panes' method, command table definition, etc) have been
+;;; executed.
+
+(defgeneric handle-frame-class-redefinition (class)
+  (:method ((class t))))
```

Comment:
i.e are we expecting that someone will change-class? if so, could we handle it in change-class method then?

Reply:
I would like this to work as robustly and unobtrusively as possible.

Imagine a user uses the mechanism in an incorrect `define-application-frame` call. Let's say something is wrong with the command table. Delayed notifications are set up, the frame class gets defined, then the user is dropped into the debugger. Maybe they undefine the frame class, define a new frame class of the same name but with a different metaclass and continue. The delayed notification would still happen despite the class having changed in the meantime. I wouldn't want the user to be dropped into the debugger a second time because of that.